Add parser for /proc/[pid]/maps

xmartinez opened the pull request:
Hi!
This adds support for parsing process memory mappings information from `/proc/[pid]/maps`.

xmartinez committed:
Add parser for /proc/[pid]/maps

xmartinez committed:
Fix build on rustc 1.13.0.

danburkert commented on src/pid/maps.rs:
````diff
@@ -0,0 +1,243 @@
+/// will be mangled by the kernel and decoded by this module as:
+///
+/// ```rust,ignore
+/// MemoryMapping (
````
These delimiters should be braces, right?

xmartinez replied:
Good catch :-) Fixed!

danburkert commented on src/pid/maps.rs:
```diff
@@ -0,0 +1,243 @@
+/// MemoryMapping (
+///    ...,
+///    path: PathBuf::from("/tmp/a\nfile"),
+///    is_deleted:true,
```
add space after colon

xmartinez replied:
Good catch! Fixed!

danburkert commented on src/pid/maps.rs:
```diff
@@ -0,0 +1,243 @@
+///
+/// See `man 5 proc`, `Linux/fs/proc/task_mmu.c`, and `Linux/fs/seq_file.c`.
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub struct MemoryMapping {
```
How do you feel about `MemoryMap` as opposed to `MemoryMapping`?  It's a bit shorter, and matches `maps` more closely.

xmartinez replied:
Sure! It does sound better! I have changed it in the updated pull request. Thanks for the suggestion!

danburkert commented on src/pid/maps.rs:
```diff
@@ -0,0 +1,243 @@
+    /// Path to the file backing this mapping (for non-anonymous mappings),
+    /// pseudo-path (such as `[stack]`, `[heap]`, or `[vdso]`) for some special
+    /// anonymous mappings, or empty path for other anonymous mappings.
+    pub path: PathBuf,
```
I think it would be better to parse out the map type into a proper enum, I think putting invalid paths into a `PathBuf` is a no-no.  Maybe could look something like this:

```
pub enum MapType {
  Anonymous,
  File(pub PathBuf),
  Stack,
  Heap,
  Vdso,
  Vvar,
  Vsyscall,
}
```

xmartinez replied:
Good point. I agree that abusing `PathBuf` for anonymous mappings pseudo-paths is not a very clean API. However, enumerating all possible pseudo-paths might be problematic, as the exact list is not documented, is architecture and config dependent, and might change (new pseudo-paths might be added) in future kernel versions (which would break this API).

I do like the idea of using an enum to distinguish between file-backed and anonymous mappings. What do you think of doing something like this?:

```
pub struct MemoryMap {
    pub range: ops::Range<usize>,
    pub is_readable: bool,
    pub is_writable: bool,
    pub is_executable: bool,
    pub is_shared: bool,
    pub type: MapType,
}

pub enum MapType {
    /// File-backed memory mapping.
    File {
        /// Offset into the file backing this mapping.
        pub offset: u64,
        /// Device containing the file backing this mapping.
        pub dev: libc::dev_t,
        /// Inode of the file backing this mapping.
        pub inode: libc::ino_t,
        /// Path to the file backing this mapping.
        pub path: PathBuf,
        /// Whether the file backing this mapping has been deleted.
        pub is_deleted: bool,
    },
    /// Anonymous memory mapping.
    Anonymous {
        /// Pseudo-path (such as `[stack]`, `[heap]`, or `[vdso]`) for some special
        /// mappings, or empty path for generic anonymous mappings.
        pub name: Vec<u8>,
    }
}
```
i.e., using a `Vec<u8>` for anonymous mappings pseudo-paths?

danburkert replied:
> However, enumerating all possible pseudo-paths might be problematic, as the exact list is not documented, is architecture and config dependent, and might change (new pseudo-paths might be added) in future kernel versions (which would break this API).

Yeah, it would be nice if the `#[non_exhaustive]` attribute was stable.  How about adding all of the known/common variants and compromising on an `Unknown(String)` variant as a fallback for everything else?  (I think `String` is OK here, since the file is guaranteed to be valid ascii, right?)

danburkert commented on src/pid/maps.rs:
```diff
@@ -0,0 +1,243 @@
+    pub path: PathBuf,
+    /// Whether the file backing this mapping has been deleted (for
+    /// non-anonymous mappings).
+    pub is_deleted: bool,
```
I can't find in the linux source where `(deleted)` is added to the file path, could you point that out to me?  The manual doesn't mention it either.

xmartinez replied:
The ` (deleted)` suffix is part of the `dcache` interface. Pathnames are generated in `/proc` files using [seq_file_path()](https://grokbit.com/torvalds/linux/ref/master/fs/seq_file.c?fn=seq_file_path&hit=1), which ends up calling [d_path()](https://grokbit.com/torvalds/linux/ref/master/fs/dcache.c?q=d_path&hit=3) to generate the path for the corresponding dentry. The latter function is the one appending ` (deleted)` if the entry has been deleted.

I have added a reference to `Linux/fs/dcache.c` in the `MemoryMap` documentation. Thanks for pointing this out! (It was indeed not obvious).

danburkert replied:
Sounds good, thanks.

danburkert commented on src/unmangle.rs:
````diff
@@ -0,0 +1,102 @@
+/// To decode only escaped newlines (leaving other escaped sequences alone):
+///
+/// ```rust,ignore
+/// let path = unmangled_path(br"a\012\040path", b"\n");
````
This doesn't look like correct usage of `br""`, since it doesn't include a delimiter, and why is it ignored?

xmartinez replied:
I am not sure I understand the issue re: `br""`. According to [the Rust Reference](https://doc.rust-lang.org/stable/reference/tokens.html#raw-byte-string-literals), this is valid syntax for raw string byte literals. What delimiter should be included?

Re: `ignore`, the `unmangle` module is not part of the crate public API, so `unmangled_path` cannot be accessed in doctests. As `rustdoc --test` tries to run doctests for private items (see https://github.com/rust-lang/rust/issues/30094), the example has to be explicitly ignored.

danburkert replied:
> I am not sure I understand the issue re: br"". According to the Rust Reference, this is valid syntax for raw string byte literals. What delimiter should be included?

Ah, so my confusion is that this instance isn't using any `#` delimiters.  Apparently that's legal, but it works the same as a non-raw literal, so it can be simplified to `b"a\012\040path"`.

> Re: ignore, the unmangle module is not part of the crate public API, so unmangled_path cannot be accessed in doctests. As rustdoc --test tries to run doctests for private items (see rust-lang/rust#30094), the example has to be explicitly ignored.

Interesting, didn't know about that limitation.  SGTM.

xmartinez replied:
I do not think `br"..."` can be simplified to `b"..."` in this case. The `r` prefix ensures that the backslash is interpreted literally (otherwise, it combines with the following b`0` byte to generate a NUL byte: https://play.rust-lang.org/?gist=2bd4e99cd3a1468149226af57fb48956&version=stable).

danburkert commented on src/unmangle.rs:
```diff
@@ -0,0 +1,102 @@
+    escaped: &'a [u8],
+}
+
+impl<'a> Iterator for UnmangledPath<'a> {
```
The iterator here is obfuscating the logic, could you just fold it into `unmangled_path` as a for loop?

xmartinez replied:
Sure! The code does indeed look more straightforward without the iterator. Thanks for the feedback :-)

danburkert reviewed (changes requested):
Thanks for the PR!  Left some specific feedback, I think the most important thing to nail down is an ergonomic public API.

xmartinez committed:
Update with PR feedback.

- Rename `MemoryMapping` as `MemoryMap`.
- Fix `MemoryMap` example.
- Refer to `Linux/fs/dache.c` in `MemoryMap` documentation (to
  document ` (deleted)` suffix appending).
- unmangle: Remove `UnmangledPath` iterator and inline decoding in
  `unmangled_path()`.

xmartinez commented:
Hi Dan!

Thanks a lot for taking the time to review this patch! (and sorry for the delay in answering, I have been sick for a couple of weeks). 

I have incorporated most of your suggested changes in the update to the PR (thanks for the feedback, the code looks much better now).

The only pending point is the `MapType {File, Anonymous}` enum change to the API (see my comment above). Could you comment on my proposal, please? (I'd rather have your feedback before changing the API).

xmartinez committed:
Change pid::maps API

- Add `MemoryMapKind` enum to encode mapping type (file-backed,
  anonymous, vdso, ...).
- Move file-backed mapping specific fields to
  `MemoryMapKind::File(FileMap)` enum.
- Add `MemoryMap.file()` convenience method.

danburkert commented:
Sounds good!  No worries about the delay, I'm not always the quickest myself :)

xmartinez commented:
Hi Dan!

I finally got around to change the API as per your suggestions. I have added a new `MemoryMapKind` enum that encodes the mapping type (file-backed, generic anonymous, vDSO, etc.).

I have also moved all fields that only make sense for file-backed mappings to a new `FileMap` struct, which is used by the `MemoryMapKind::File` enum variant.

I have added a `file() -> Option<&FileMap>` method to the `MemoryMap` struct to make client code that is only interested in file-backed mappings somewhat simpler (it avoids the necessity of using pattern matching in some use cases).

Apart from that, I have also extracted device number parsing onto its own function `parse_dev()` and the code now uses the device number to determine whether the mapping is file-backed (the previous approach used the inode number, which should also work, but comparing to the null device number will definitely work).

Let me know if there are any other changes that might improve the code. If not, I can rebase and squash the pull request branch (so it can be merged as a single logical commit).

ice1000 commented:
Any update?
